src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ContentLengthReadStream.cs

```diff
@@ -133,6 +133,32 @@ private void Finish()
                 _connection.ReturnConnectionToPool();
                 _connection = null;
             }
+
+            public override async Task DrainAsync(int maxDrainBytes)
+            {
+                Debug.Assert(_connection != null);
+                Debug.Assert(_contentBytesRemaining > 0);
+
+                if (_contentBytesRemaining > (ulong)maxDrainBytes)
+                {
+                    return;
+                }
```

Comment:
Why do we want this check here?  Seems like we should only have it before doing the FillAsync; if we already have the remaining content buffered, we may as well "consume" it and finish so the connection can be pooled.  Maybe it doesn't matter in practice, if maxDrainBytes is larger than the connection buffer size, but still seems odd to have this here.  Or am I missing something?

Reply:
No, you're not missing anything.  I did in fact do something like that for chunked handling -- we always process the full buffer before checking if we've exceeded the drain limit.

My thinking here was exactly what you say -- in practice, maxDrainBytes will always (usually?) be greater than the connection buffer size.  But that's sort of lazy and may not be actually true.  I'll change the code.